src/Compilers/Core/CodeAnalysisTest/Diagnostics/SystemLanguageAnalyzer.cs

```diff
@@ -128,4 +132,118 @@ void Report(CompilationAnalysisContext context, IFieldSymbol field, DiagnosticDe
             context.ReportDiagnostic(Diagnostic.Create(descriptor, field.Locations.FirstOrDefault()));
         }
     }
+
+    /// <summary>Analyzer used to identify local variables that could be declared Const.</summary>
+    public class LocalCouldBeConstAnalyzer : DiagnosticAnalyzer
+    {
+        private const string SystemCategory = "System";
+
+        public static readonly DiagnosticDescriptor LocalCouldBeConstDescriptor = new DiagnosticDescriptor(
+            "LocalCouldBeReadOnly",
+            "Local Could Be Const",
+            "Local variable is never modified and so could be const.",
+            SystemCategory,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true);
+
+        /// <summary>Gets the set of supported diagnostic descriptors from this analyzer.</summary>
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get { return ImmutableArray.Create(LocalCouldBeConstDescriptor); }
+        }
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationBlockStartAction(
+                (operationBlockContext) =>
+                {
+                    IMethodSymbol containingMethod = operationBlockContext.OwningSymbol as IMethodSymbol;
+
+                    if (containingMethod != null)
+                    {
+                        HashSet<ILocalSymbol> mightBecomeConstLocals = new HashSet<ILocalSymbol>();
+                        HashSet<ILocalSymbol> assignedToLocals = new HashSet<ILocalSymbol>();
+
+                        operationBlockContext.RegisterOperationAction(
+                           (operationContext) =>
+                           {
+                               IAssignmentExpression assignment = (IAssignmentExpression)operationContext.Operation;
+                               AssignTo(assignment.Target, assignedToLocals, mightBecomeConstLocals);
+                           },
+                           OperationKind.AssignmentExpression,
```

Comment:
Unrelated to this PR, but I really hope someday we could have API overloads that takes the type of the node/operation instead of syntax kind/operation kind. Otherwise maintaining such analyzers will be extremely tedious as and when a node type/operation type supports a new kind.